foubus.py

```diff
@@ -42,411 +53,552 @@
+        trip_date = entity.trip_update.trip.start_date
+
+        if (tt["trip_id"] == trip_id).any():
             logging.info(
                 "trip_update for %s: %s: %d stop_time_update",
-                entity.trip_update.trip.trip_id,
+                trip_id,
                 text_format.MessageToString(entity.trip_update.trip, as_one_line=True),
                 len(entity.trip_update.stop_time_update),
             )
             last_stop_sequence = None
+
             for stu in entity.trip_update.stop_time_update:
-                # TODO: implement delay propagation
-                # https://gtfs.org/documentation/realtime/feed-entities/trip-updates/#:~:text=If%20one%20or%20more%20stops%20are%20missing
-                assert (
-                    last_stop_sequence is None
-                    or last_stop_sequence + 1 == stu.stop_sequence
-                ), text_format.MessageToString(stu, as_one_line=True)
-
-                if (
-                    stu.schedule_relationship
-                    != gtfs_realtime_pb2.TripUpdate.StopTimeUpdate.ScheduleRelationship.SCHEDULED
-                ):
+                # Validate consecutive stop sequences
+                if last_stop_sequence is not None and last_stop_sequence + 1 != stu.stop_sequence:
+                    logging.warning(
+                        "Non-consecutive stop sequence: %s",
+                        text_format.MessageToString(stu, as_one_line=True),
+                    )
+                last_stop_sequence = stu.stop_sequence
+
+                # Only apply updates for scheduled stops
+                if stu.schedule_relationship != stu.ScheduleRelationship.SCHEDULED:
                     continue
 
-                row = tt[
-                    (tt["trip_id"] == entity.trip_update.trip.trip_id)
-                    & (tt["date"] == entity.trip_update.trip.start_date)
+                mask = (
+                    (tt["trip_id"] == trip_id)
+                    & (tt["date"] == trip_date)
                     & (tt["stop_sequence"] == stu.stop_sequence)
                     & (tt["stop_id"] == stu.stop_id)
+                )
+                row_df = tt[mask]
+                if row_df.empty:
+                    continue
+
+                if not stu.departure.time:
+                    logging.warning(
+                        "No departure time for trip: %s, stop_time_update: %s",
+                        text_format.MessageToString(entity.trip_update.trip, as_one_line=True),
+                        text_format.MessageToString(stu, as_one_line=True),
+                    )
+                    continue
+
+                departure_dt = datetime.datetime.fromtimestamp(stu.departure.time)
+                tt.loc[mask, ["realtime", "leave_in"]] = [
+                    True,
+                    departure_dt - now,
                 ]
-                if not row.empty:
-                    assert len(row) == 1, row
```

Comment:
Here too, please keep the assert. It's a fundamental invariant of the input data, and if it ever seems to be false, I want a crash and will investigate.